python/cudf/cudf/core/column/column.py

```diff
@@ -1161,55 +1091,71 @@ def as_column(arbitrary, nan_as_null=True, dtype=None, name=None):
+        children = (
+            build_column(obuf, dtype="int32"),
+            build_column(sbuf, dtype="int8"),
+        )
+        data = build_column(
+            data=None, dtype="object", mask=nbuf, children=children
+        )
+        data._nvstrings = arbitrary
+
+    elif isinstance(arbitrary, Buffer):
+        if dtype is None:
+            raise TypeError(f"dtype cannot be None if 'arbitrary' is a Buffer")
+        data = build_column(arbitrary, dtype=dtype)
 
     elif cuda.devicearray.is_cuda_ndarray(arbitrary):
-        data = as_column(Buffer(arbitrary))
+        data = as_column(Buffer(arbitrary), dtype=arbitrary.dtype)
         if (
             data.dtype in [np.float16, np.float32, np.float64]
             and arbitrary.size > 0
         ):
             if nan_as_null:
                 mask = libcudf.unaryops.nans_to_nulls(data)
-                data = data.set_mask(mask)
+                data.mask = mask
 
         elif data.dtype.kind == "M":
-            null = cudf.core.column.column_empty_like(
-                data, masked=True, newsize=1
-            )
+            null = column_empty_like(data, masked=True, newsize=1)
             col = libcudf.replace.replace(
-                as_column(Buffer(arbitrary)),
+                as_column(Buffer(arbitrary), dtype=arbitrary.dtype),
                 as_column(
-                    Buffer(np.array([np.datetime64("NaT")], dtype=data.dtype))
+                    Buffer(np.array([np.datetime64("NaT")], dtype=data.dtype)),
+                    dtype=arbitrary.dtype,
                 ),
                 null,
             )
             data = datetime.DatetimeColumn(
-                data=Buffer(arbitrary), mask=col.mask, dtype=data.dtype
+                data=Buffer(arbitrary),
+                dtype=data.dtype,
+                mask=col.mask,
+                name=name,
             )
 
     elif hasattr(arbitrary, "__cuda_array_interface__"):
         desc = arbitrary.__cuda_array_interface__
         data = _data_from_cuda_array_interface_desc(desc)
         mask = _mask_from_cuda_array_interface_desc(desc)
+        dtype = np.dtype(desc["typestr"])
+        nelem = desc["shape"][0]
+        col = build_column(data, dtype=dtype, mask=mask, name=name)
 
-        if mask is not None:
-            nelem = len(data.mem)
-            nnz = libcudf.cudf.count_nonzero_mask(mask.mem, size=nelem)
-            null_count = nelem - nnz
-        else:
-            null_count = 0
-        col = build_column(
-            data, dtype=data.dtype, mask=mask, name=name, null_count=null_count
-        )
         # Keep a reference to `arbitrary` with the underlying
-        # RMM device array, so that the memory isn't freed out
+        # Buffer, so that the memory isn't freed out
         # from under us
-        col.data.mem._obj = arbitrary
+        col.data._obj = arbitrary
         return col
 
     elif isinstance(arbitrary, np.ndarray):
```

Comment:
Should just handle `__array_interface__` instead of `np.ndarray` directly.

Reply:
Hmm, we're doing some ndarray specific stuff here (fixing contiguity)..

Reply:
Shouldn't we do that for anything with `__array_interface__` that is non-contiguous?

Reply:
Will raise an issue after this is merged.